
## Required data
- Raw data is available at EGA with accesion X
- Intermediate files can be downloaded from Zenodo: 

## Iso-Seq analysis with IsoQuant

### Software
| Tool  | Version | Link |
| ----- | ------- | ---- | 
| Isoseq3 | 3.4.0 | https://github.com/PacificBiosciences/IsoSeq   |
| Minimap2  | v2.17-r941 | https://github.com/lh3/minimap2  |
| SAMtools  | 1.15 | https://github.com/samtools/samtools |
| IsoQuant | 3.1.2 | https://github.com/ablab/IsoQuant |

### Scripts
1. [CCS](ccs.sh) - creates circular consensus sequence of the raw reads
2. [Isoseq](isoseq.sh) - primer removal, isoseq refine and mapping to reference genome
3. [IsoQuant](isoquant.sh) - isoform classification with IsoQuant


## Iso-Seq analysis with SQANTI3

## Software
| Tool  | Version | Link |
| ----- | ------- | ---- | 
| Isoseq3 | 3.4.0 | https://github.com/PacificBiosciences/IsoSeq   |
| Minimap2  | v2.17-r941 | https://github.com/lh3/minimap2  |
| SAMtools  | 1.15 | https://github.com/samtools/samtools |
| cDNAcupcake | |  |
| SQANTI3 | | |

## Iso-Seq analysis with TALON


