## Data

[BioMart](biomart) contains information downloaded from Ensembl
- `proteincoding_genes.txt` and `proteincoding_transcripts.txt` are used to create the transcript length plot in Figure 2A
- `transcript_length.txt` is used to correlate the transcrript length with the number of isoforms in Figure S1B

[Proteomics](proteomics)
- `Human_Hexamer.tsv` and `Human_logitModel.RData` are necessary to run CPAT
- `gene.gtf` and `gene_peptide.gtf` files are used to create Figure 5

[Reference genome](ref_genome)
- 'gene_gencode.gtf' files are used to create Figures 5 and 6

[Retina specific exons](retina_specific_exons) are used to compare our results to previous studies
- `short_exons.bed` and `long_exons.bed` are retina-enriched exons derived from Ciampi at al. (2022)
- `Musashi_hg38.bed` contains photoreceptor-enriched identified by Murphi et al. (2016)

[RetNet](retnet) contains a list of unique IRD-associated genes downloaded from [RetNet](https://web.sph.uth.edu/RetNet/) on 19 May 2022

[Transcriptomics](transcriptomics)
- `gene.gtf` files are used to create Figure 4
- `primers.fasta` is used for PacBio long-read RNA-sequencing analysis

### Additional bigger data files are available at:
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10371435.svg)](https://doi.org/10.5281/zenodo.10371435)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10371395.svg)](https://doi.org/10.5281/zenodo.10371395)

